File: Graphs/graph-bellman-ford.js
// Bellman-Ford Algorithm
// Time: O(V * E)
// Space: O(V)

function bellmanFord(edges, V, start) {
    const dist = Array(V).fill(Infinity);
    dist[start] = 0;

    for (let i = 1; i <= V - 1; i++) {
        for (const [u, v, w] of edges) {
            if (dist[u] !== Infinity && dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
            }
        }
    }

    // Check for negative-weight cycles
    for (const [u, v, w] of edges) {
        if (dist[u] !== Infinity && dist[u] + w < dist[v]) {
            throw new Error("Graph contains negative weight cycle");
        }
    }

    return dist;
}

File: Graphs/graph-bellman-ford.java
// Bellman-Ford Algorithm
// Time: O(V * E)
// Space: O(V)

import java.util.*;

public class BellmanFord {
    static class Edge {
        int src, dest, weight;
        Edge(int s, int d, int w) {
            src = s; dest = d; weight = w;
        }
    }

    public static int[] bellmanFord(int V, List<Edge> edges, int start) {
        int[] dist = new int[V];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[start] = 0;

        for (int i = 1; i <= V - 1; i++) {
            for (Edge edge : edges) {
                if (dist[edge.src] != Integer.MAX_VALUE &&
                    dist[edge.src] + edge.weight < dist[edge.dest]) {
                    dist[edge.dest] = dist[edge.src] + edge.weight;
                }
            }
        }

        // Check for negative-weight cycles
        for (Edge edge : edges) {
            if (dist[edge.src] != Integer.MAX_VALUE &&
                dist[edge.src] + edge.weight < dist[edge.dest]) {
                throw new RuntimeException("Graph contains negative weight cycle");
            }
        }

        return dist;
    }
}
